src/object.d

```diff
@@ -482,6 +482,167 @@ unittest
     assert(a1 != a2);
 }
 
+/**
+Destruct the given object. Correctly call its destructor or finalizer so it no longer
+references any other objects. This $(I does not) reset the memory to init state. After
+this call succeeds, the object memory is in an invalid state, and should not be referenced.
+It does $(I not) initiate a GC cycle or free any GC memory.
+*/
+void destruct(T)(ref T obj) if (is(T == struct))
+{
+    _destructRecurse(obj);
+}
+
+///
+void destruct(T)(T obj) if (is(T == class))
+{
+    static if(__traits(getLinkage, T) == "C++")
+        obj.__xdtor();
+    else
+        rt_finalize(cast(void*)obj); // can destruct without init?
+}
+
+///
+void destruct(T)(T obj) if (is(T == interface))
+{
+    destruct(cast(Object)obj);
+}
+
+///
+void destruct(T : U[n], U, size_t n)(ref T obj) if (!is(T == struct))
```

Comment:
I guess a struct could `alias this` an array, and then it would incorrectly be accepted by the function?

Reply:
Yeah, I tried that and you're correct. But it applies to classes too.

Reply:
Bug in `destroy`?

Reply:
It also applies to unions. The full condition should be "is not an aggregate type". That concept should arguably be exposed via `__traits` since it's already visible within DMD and it's so commonly needed in template conditions.

https://run.dlang.io/is/XlUagV
```d
void destruct(T : U[n], U, size_t n)(T t)
{
}

union U
{
    uint[0] a;
    alias a this;
}

void main()
{
    destruct(U.init);
}
```

Reply:
One could argue that unions should be disallowed to have an `alias this`.

Reply:
Why?

Reply:
Because union treatment must be explicit, as they forego usual type semantics. Subtyping a tagless union just makes no sense. Although come to think of it, this has little to do with this PR, I'd rather take it elsewhere.